### PR TITLE
Refactor history cache to its own package

### PR DESCRIPTION
**Pull request by mindaugasrukas**
**What changed?**
Refactored history cache to its own package. Together Refactored DeleteManager to its package to avoid circular imports.

**Why?**
Code preparation for: https://github.com/temporalio/temporal/pull/3594

**How did you test it?**
Unit tests, CI, `make bins`

**Potential risks**
Low risk as unit tests and builds are passing; this is just package renaming—no logic change.

**Is hotfix candidate?**
None

### CODE REVIEW

**Review by yycptt — COMMENTED** [review]
Plz... can it be called workflow cache, instead of history cache as it's caching workflow mutable state instead of workflow history events? All the variable names are also workflowCache. And I remember there're some refactor before (either in temporal or cadence repo) to change the name from historyCache to workflow cache.

Is it possible to create the new package under `workflow`? Like `service/history/workflow/cache`?

File: service/history/api/consistency_checker.go [hunk 1]
```diff
@@ -37,6 +37,7 @@ import (
 	"go.temporal.io/server/common/namespace"
 	"go.temporal.io/server/common/persistence"
 	"go.temporal.io/server/common/persistence/versionhistory"
+	historyCache "go.temporal.io/server/service/history/cache"
```
Comment (alexshtin) [inline service/history/api/consistency_checker.go:40]: I think `hcache` is more canonical go import prefix for `history/cache` package.

File: service/history/cache/fx.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	"go.temporal.io/server/service/history/shard"
+)
+
+var Module = fx.Options(
+	fx.Provide(NewCacheFnProvider),
+)
+
+// NewCacheFnProvider provide a NewCacheFn that can be used to create new workflow cache.
+func NewCacheFnProvider() NewCacheFn {
+	return func(shard shard.Context) Cache {
```
Comment (alexshtin) [inline service/history/cache/fx.go:39]: I understand this is moved from another file, do you have an idea why `func` is used here?

**Review by alexshtin — APPROVED** [review]

**Comment by alexshtin** [comment]
> Sounds good. I'm just curious about, why `workflow` is under history (`service/history/workflow`). Is it `HistoryWorkflow` and `HistoryWorkflowCache`?

`history` here is related to `service` -- it is "history service". And all packages under it are used by this `history service` only.

File: service/history/cache/fx.go (hunk 2 above)
Comment (mindaugasrukas) [reply service/history/cache/fx.go:39]: I don't know, but it sounds like an implementation of lazy loading. And AFAIK, Uber/Fx doesn't support lazy loading.

File: service/history/api/consistency_checker.go (hunk 1 above)
Comment (mindaugasrukas) [reply service/history/api/consistency_checker.go:40]: moving and renaming to `wcache`.

File: service/history/deletemanager/delete_manager.go [hunk 3]
```diff
@@ -53,32 +55,32 @@ type (
...
+			ms workflow.MutableState,
 			archiveIfEnabled bool,
 			stage *tasks.DeleteWorkflowExecutionStage,
 		) error
 	}
 
 	DeleteManagerImpl struct {
 		shard          shard.Context
-		historyCache   Cache
+		workflowCache  wcache.Cache
```
Comment (mindaugasrukas) [inline service/history/deletemanager/delete_manager.go:83]: fyi

File: service/history/ndc/activity_replicator.go [hunk 4]
```diff
@@ -68,20 +66,20 @@ type (
 	}
 
 	ActivityReplicatorImpl struct {
-		historyCache    workflow.Cache
+		workflowCache   wcache.Cache
```
Comment (mindaugasrukas) [inline service/history/ndc/activity_replicator.go:69]: fyi

File: service/history/ndc/history_replicator.go [hunk 5]
```diff
@@ -113,7 +113,7 @@ type (
 		historySerializer serialization.Serializer
 		metricsHandler    metrics.MetricsHandler
 		namespaceRegistry namespace.Registry
-		historyCache      workflow.Cache
+		workflowCache     wcache.Cache
```
Comment (mindaugasrukas) [inline service/history/ndc/history_replicator.go:116]: fyi

File: service/history/ndc/history_replicator_test.go [hunk 6]
```diff
@@ -66,7 +67,7 @@ type (
 		controller            *gomock.Controller
 		mockShard             *shard.ContextTest
 		mockEventCache        *events.MockCache
-		mockHistoryCache      *workflow.MockCache
+		mockWorkflowCache     *wcache.MockCache
```
Comment (mindaugasrukas) [inline service/history/ndc/history_replicator_test.go:70]: fyi

File: service/history/ndc/transaction_manager.go [hunk 7]
```diff
@@ -154,7 +154,7 @@ type (
 	transactionMgrImpl struct {
 		shard             shard.Context
 		namespaceRegistry namespace.Registry
-		historyCache      workflow.Cache
+		workflowCache     wcache.Cache
```
Comment (mindaugasrukas) [inline service/history/ndc/transaction_manager.go:157]: fyi

File: service/history/ndc/workflow_resetter.go [hunk 8]
```diff
@@ -82,7 +82,7 @@ type (
 		namespaceRegistry namespace.Registry
 		clusterMetadata   cluster.Metadata
 		executionMgr      persistence.ExecutionManager
-		historyCache      workflow.Cache
+		workflowCache     wcache.Cache
```
Comment (mindaugasrukas) [inline service/history/ndc/workflow_resetter.go:85]: fyi

File: service/history/queueProcessorBase.go [hunk 9]
```diff
@@ -89,7 +89,7 @@ func newQueueProcessorBase(
 	options *QueueProcessorOptions,
 	queueProcessor common.Daemon,
 	queueAckMgr queueAckMgr,
-	historyCache workflow.Cache,
+	wcache wcache.Cache,
```
Comment (mindaugasrukas) [inline service/history/queueProcessorBase.go:92]: fyi

File: service/history/workflow/cache/cache.go [hunk 10]
```diff
@@ -22,9 +22,9 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 // THE SOFTWARE.
 
-//go:generate mockgen -copyright_file ../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination cache_mock.go
+//go:generate mockgen -copyright_file ../../../../LICENSE -package $GOPACKAGE -source $GOFILE -destination cache_mock.go
```
Comment (mindaugasrukas) [inline service/history/workflow/cache/cache.go:25]: fyi

File: service/history/workflow/cache/cache_test.go [hunk 11]
```diff
@@ -45,10 +45,11 @@ import (
 	"go.temporal.io/server/common/persistence"
 	"go.temporal.io/server/service/history/shard"
 	"go.temporal.io/server/service/history/tests"
+	"go.temporal.io/server/service/history/workflow"
 )
 
 type (
-	historyCacheSuite struct {
+	workflowCacheSuite struct {
```
Comment (mindaugasrukas) [inline service/history/workflow/cache/cache_test.go:52]: fyi

**Review by yycptt — APPROVED** [review]

File: service/history/workflowTaskHandlerCallbacks_test.go [hunk 12]
```diff
@@ -105,7 +106,7 @@ func (s *WorkflowTaskHandlerCallbackSuite) SetupTest() {
 	s.mockEventsCache.EXPECT().PutEvent(gomock.Any(), gomock.Any()).AnyTimes()
 	s.logger = mockShard.GetLogger()
 
-	historyCache := workflow.NewCache(mockShard)
+	workflowCache := wcache.NewCache(mockShard)
```
Comment (mindaugasrukas) [inline service/history/workflowTaskHandlerCallbacks_test.go:109]: fyi

**Pull request merged**